src/primitives/transaction.h

```diff
@@ -928,12 +964,12 @@ class CTransaction
 
     friend bool operator==(const CTransaction& a, const CTransaction& b)
     {
-        return a.hash == b.hash;
+        return a.wtxid.hash == b.wtxid.hash;
```

Comment:
This now means that two transactions are equal if their effecting data is equal. Two questions:
(1) Do we really need transaction equality checking? Could we get by with explicitly dereferencing to the wtxid/txid and doing equality comparisons on those explicitly?
(2) If we really need this, are we sure that allowing two transactions with differing authorizing data to be considered equal can't cause problems?

Reply:
(This is blocking, the rest of my comments from this review are nb).

Reply:
> (1) Do we really need transaction equality checking? Could we get by with explicitly dereferencing to the wtxid/txid and doing equality comparisons on those explicitly?

If we want to store `CTransactions` in any C++ collection that relies on equality (e.g. hashmap), this method must be implemented. IDK if we do, but that's the most common reason for these methods being implemented.

> (2) If we really need this, are we sure that allowing two transactions with differing authorizing data to be considered equal can't cause problems?

This is one of the things that I was concerned about getting right. The existing assumption throughout the codebase, that this PR maintains, is that the node does not distinguish between different transactions that have identical effects. All of the places where we store actual `CTransaction`s (not just references to them), we do so only after validating the transaction (either via `AcceptToMemoryPool` or `ConnectBlock`), at which point the authorizing data is immaterial.

Upstream Bitcoin Core implements it this way as well, and I defer in part to them as the progenitor of our codebase. See [here](https://github.com/bitcoin/bitcoin/pull/18044#discussion_r460704950) for example of their discussion on this point. They implemented this check by txid, and continue to do so for SegWit transactions (and did not change to using wtxid for equality checks):

https://github.com/bitcoin/bitcoin/blob/40a9037a1b5d990637d7f5009fc0c39628ed2c05/src/primitives/transaction.h#L320-L323

Reply:
Okay. Ideally I think we'd use a wrapper type that defines equality separately for collection storage rather than having a single global definition, but I'll defer to the fact that upstream thinks this is okay.

Reply:
I'm tempted to just remove the equality checks and see what fails to compile. :)

Reply:
This almost works: https://github.com/str4d/zcash/pull/23

The only place we then use transaction equality is in gmock matchers like https://github.com/zcash/zcash/blob/master/src/wallet/gtest/test_wallet.cpp#L1575